bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/sitemap/ProductsSitemapGenerator.java

```diff
@@ -0,0 +1,166 @@
+
+    @ObjectClassDefinition(name = "CIF Product Sitemap Generator")
+    @interface Configuration {
+
+        @AttributeDefinition(
+            name = "Pagination Size",
+            description = "The number of products to query from the commerce backend per iteration.")
+        int pageSize() default 10;
+    }
+
+    static final String PN_NEXT_PRODUCT = "nextProduct";
+    static final String PN_NEXT_PAGE = "nextPage";
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ProductsSitemapGenerator.class);
+
+    @Reference
+    private UrlProvider urlProvider;
+    @Reference
+    private SitemapLinkExternalizer externalizer;
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL, policyOption = ReferencePolicyOption.GREEDY)
+    private SitemapProductFilter productFilter;
+
+    private int pageSize;
+
+    @Activate
+    protected void activate(Configuration configuration) {
+        this.pageSize = configuration.pageSize();
+    }
+
+    @Override
+    public Set<String> getNames(Resource sitemapRoot) {
+        Page page = sitemapRoot.adaptTo(Page.class);
+        Page specificPage = page != null ? SiteNavigation.getProductPage(page) : null;
+        return specificPage != null && specificPage.getPath().equals(page.getPath())
+            ? Collections.singleton(SitemapService.DEFAULT_SITEMAP_NAME)
+            : Collections.emptySet();
+    }
+
+    @Override
+    public void generate(Resource sitemapRoot, String name, Sitemap sitemap, SitemapGenerator.Context context) throws SitemapException {
+        MagentoGraphqlClient graphql = sitemapRoot.adaptTo(MagentoGraphqlClient.class);
+        Page productPage = sitemapRoot.adaptTo(Page.class);
+
+        if (graphql == null || productPage == null) {
+            throw new SitemapException("Failed to build product sitemap at: " + sitemapRoot.getPath());
+        }
+
+        int currentIndex = context.getProperty(PN_NEXT_PRODUCT, 0);
+        int currentPageIndex = context.getProperty(PN_NEXT_PAGE, 1);
+        int maxPages = Integer.MAX_VALUE;
+        // parameter map to be reused while iterating
+        UrlProvider.ParamsBuilder paramsBuilder = new UrlProvider.ParamsBuilder()
+            .page(externalizer.externalize(sitemapRoot));
+
+        while (currentPageIndex <= maxPages) {
+            String query = Operations.query(productsQueryFor(currentPageIndex, pageSize)).toString();
+            GraphqlResponse<Query, Error> resp = graphql.execute(query);
+
+            if (resp.getErrors() != null && resp.getErrors().size() > 0) {
+                SitemapException ex = new SitemapException("Failed to execute graphql query.");
+                resp.getErrors().forEach(error -> ex.addSuppressed(new Exception(error.getMessage())));
+                throw ex;
+            }
+
+            Products products = resp.getData().getProducts();
+            List<ProductInterface> items = products.getItems();
+            maxPages = products.getTotalCount() / pageSize;
```

Comment:
This could be a one liner with `Math.ceil`.

Reply:
this would require a cast from either of the operations arguments to double and returns a double which has again to be casted. 
``` 
int maxPages = (int) Math.ceil((double) products.getTotalCount() / pageSize)
```

Let's keep it as it is.